Add aml_pipeline (#15)
eProsima-Private/AML-Tools: docs/rst/tutorials/aml_pipeline.rst

.. include:: /rst/exports/roles.include

.. _tutorials_aml_pipeline:

> AML Pipeline

Reviewer: I'd remove this title and replace it with "Running a binary classifier with MNIST". The tutorial does not have to be name "AML Pipeline" as this section was just created to build the skeleton of the documentation. Also, change the name of the file to ``binary_classifier_mnist.rst``.

> ############

Reviewer: Add a more detailed description of this tutorial. Something like "This tutorial covers...".

You can request this Jupyter notebook and AML libraries in :ref:`Get access to AML Toolkit <get_access_toolkit>`.

> Running a binary classifer with MNIST

Reviewer: ```suggestion
Running a binary classifier with MNIST
```

*************************************

Load AML libraries
==================

Most functionality to work with AML can be found in two packages:

> - ``aml_engine.amlSimpleLibrary`` contains functions and classes to create and manipulate AML models,

Reviewer: The list is not rendered correctly. Add a line break.

```suggestion

- ``aml_engine.amlSimpleLibrary`` contains functions and classes to create and manipulate AML models.
```

- ``aml_engine.amlAuxiliaryLibrary`` contains functions and classes specific for training.

.. code-block:: python

    from aml_engine import amlSimpleLibrary as sc
    from aml_engine import amlAuxiliaryLibrary as ql

Load MNIST dataset
==================

The original dataset files:

- ``t10k-images-idx3-ubyte``
- ``t10k-labels-idx1-ubyte``
- ``train-images-idx3-ubyte``
- ``train-labels-idx1-ubyte``

> The dataset can be downloaded from:
> http://yann.lecun.com/exdb/mnist/
> and
> https://huggingface.co/datasets/ylecun/mnist

Reviewer: ```suggestion
The dataset can be downloaded from `Yann Lecun MNIST database <http://yann.lecun.com/exdb/mnist/>`__ or `Huggingface <https://huggingface.co/datasets/ylecun/mnist>`__.
```

.. code-block:: python

    from MNIST import mnistGenerator

The ``mnistGenerator`` searches and loads the datasets and provides functionality to load individual digits.

Embedding constants
===================

> A simple embedding for black and white images:

Reviewer: Same as before, add a break line before lists to render them correctly. 

```suggestion
A simple embedding for black and white images:

```

- constants for each black pixel `B(i,j)` (height x width),
- constants for each white pixel `W(i,j)` (height x width),
- constants for each label (10 if using all digits).

For instance, with 2x2 images the embedding set would be
``[B(1,1), B(1,2), B(2,1), B(2,2), W(1,1), W(1,2), W(2,1), W(2,2), label(1), label(2), ...]``

> ### Embedding function

Reviewer: ```suggestion
Embedding function
==================
```

.. code-block:: python

    def mnistDigit(targetDigit, size):
        d, label, i = MNIST_GENERATOR.getNextDigit(targetDigit, False)
        result = digitToConstants(d, size)
        return sc.amlset(result)

    def mnistOtherDigit(targetDigit, size):
        d, label, i = MNIST_GENERATOR.getNextDigit(targetDigit, True)
        result = digitToConstants(d, size)
        return sc.amlset(result)

Training
========

Training parameters
^^^^^^^^^^^^^^^^^^^

.. code-block:: python

    # Training iterations
    iterations = 1000 # Maximum number of mini-batch iterations

    # Problem parameters
    validation_size = 10000
    MNIST_GENERATOR = mnistGenerator(validation_size)
    gridSideLength = 28  # side of images in the MNIST dataset
    embedding_constants = {*range(0, 2 * gridSideLength * gridSideLength)}
    targetDigit = 0  # 0 ... 9

    balance = 9
    pBatchSize = 100
    nBatchSize = 100 * balance
    maxPTrainingExamples = 6000
    maxNTrainingExamples = 54000

Notice that we are picking one target digit.
The model will discriminate between that digit (positive class), and the rest.

> It is also possible to slightly modify this script for the multiclass problem.

Reviewer: I'd move the description before the code.

> It is also possible to slightly modify this script for the multiclass problem.

Reviewer: Move this sentence to another section at the end of this tutorial called "Next Steps" and add a link there to another tutorial named "Extending the MNIST binary classifier to a multiclass classifier". Create the file for this tutorial and add it to the index. Then on the file just add the title and a warning with a "Work in progress" statement.

Author: No sé si es una buena idea. Es sencillo extender el código a multiclase, pero la inferencia se vuelve mucho más complicada (a nivel estadístico, no de AML)

Reviewer: Let's skip it for now then

Model and batch trainer initialisation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Initialise the ``model`` object that contains:

Reviewer: Line break

```suggestion
Initialise the ``model`` object that contains:

```

      G 7 atoms 94
      G 8 atoms 54
      G 9 atoms 22
      G 10 atoms 7
      G 11 atoms 2
      G 12 atoms 2
    L spectrum:
      L 1 atoms 452
      L 2 atoms 8
      L 3 atoms 14
      L 4 atoms 28
      L 5 atoms 46
      L 6 atoms 38
      L 7 atoms 22
      L 8 atoms 25
      L 9 atoms 18
      L 10 atoms 20
      L 11 atoms 17
      L 12 atoms 19
      L 13 atoms 20
      L 14 atoms 19
      L 15 atoms 7
      L 16 atoms 17
      L 17 atoms 17
      L 18 atoms 21
      L 19 atoms 18
      L 20 atoms 15
      L 21 atoms 19
      L 22 atoms 8
      L 23 atoms 11
      L 24 atoms 9
      L 25 atoms 8
      L 26 atoms 9
      L 27 atoms 5
      L 28 atoms 6
      L 29 atoms 4
      L 30 atoms 3
      L 31 atoms 3
      L 32 atoms 3
      L 33 atoms 2
      L 34 atoms 3
      L 35 atoms 1
      L 36 atoms 2
      L 37 atoms 3
      L 39 atoms 1
      L 40 atoms 1
      L 41 atoms 2
      L 42 atoms 1
      L 44 atoms 2
    fraction of negative indicators:  0.14285714285714285 , union model fraction:  15
    BATCH: 948 batchSize( 100 , 900              ------------- master:  model size 9977

    ... following iterations ...

Postprocessing
==============

> Here you can explore the atomization in the last model: `model.atomization`

Reviewer: ```suggestion
Here you can explore the atomization in the last model: `model.atomization`.
```

> But also the cumulative model that has been built combining the atomizations across the whole training process: `batchLearner.unionModel`

Reviewer: ```suggestion
But also the cumulative model that has been built combining the atomizations across the whole training process: `batchLearner.unionModel`.
```

.. code-block:: python

    print(f"{cGreen}Union model size:{cReset} {len(batchLearner.unionModel)}")
    print(f"{cGreen}Size spectrum:{cReset}")
    sc.printLSpectrum(batchLearner.unionModel)

    print(f"{cGreen}Some random atoms:{cReset}")
    import random
    for at in random.choices(model.atomization, k=10):
        print(at.ucs)

    batchLearner.unionModel.sort(reverse=True,key=lambda at: len(at.ucs))
    print(f"{cGreen}Largest atom{cReset}")
    print(batchLearner.unionModel[0].ucs)

> This displays the following information::

Reviewer: I'd explain more about this information, i.e. what can we deduce from it or how could we interpret it to check if everything looks good.

Plotting results
^^^^^^^^^^^^^^^^

For classification problems with square images the viewer module provides some functions to plot atoms.
It simply maps back constant indices to their position and color on an image.
It then uses ``matplotlib`` to output the results.

.. code-block:: python

    import viewer as av

    theViewer = av.atomViewer_bw_sq_image(
        cmanager=model.cmanager,
        min_size=2,
        scale=10,
        atoms_per_board=[10, 10],
    )
    theViewer.save(f"RESULTS/MNIST_IMG_{targetDigit}_{i}", batchLearner.unionModel)
    theViewer.save(f"RESULTS/MNIST_IMG_model_{targetDigit}_{i}", model.atomization)

As an example, below we show 100 atoms in ``model.atomization``.
Black and white pixels indicate that the constant is present in the atom; red, that both constants are present; and gray, that none is present.

.. image:: /rst/figures/tutorials/example_atoms_mnist.png
   :width: 400
   :alt: A hundred atoms from the atomization of MNIST.

>    :align: center

Reviewer: I'm missing another section for inference, I mean, given an image from MNIST dataset not used for training, infer if it belongs to that class or not.
